Add test for timeout parameter

jacquelinekay opened the pull request:
Connects to ros2/ros2#73

This test could be prone to flakiness due to scheduling nondeterminism, hence the extremely high tolerance. If we'd prefer to avoid flaky tests, I'm fine with declining this change, but I posted it to prove the correctness of my timeout PRs.

dirk-thomas commented:
This does not add any test yet only an executable.

jacquelinekay commented:
Alright, now I have actually added the nosetest in ament.

dirk-thomas commented on test_communication/test/test_timeout_subscriber.cpp:
```diff
@@ -0,0 +1,156 @@
+  auto subscriber = node->create_subscription<T>(
+    std::string("test_message_") + message_type, expected_messages.size(), callback);
+  return subscriber;
+}
```
Why all this copied complexity if the test is not actually ever running a publisher or receives a message?

dirk-thomas commented on test_communication/test/test_timeout_subscriber.cpp:
```diff
@@ -0,0 +1,156 @@
+  auto tolerance = std::chrono::milliseconds(10);
+  for (size_t i = 0; i < num_cycles; ++i) {
+    auto nonblocking_start = std::chrono::steady_clock::now();
+    executor.spin_node_once(node, std::chrono::milliseconds::zero());
```
Is this `spin_node_once` call supposed to sleep for 100 ms when the timeout parameter is zero?

jacquelinekay replied:
No. When the timeout parameter is zero, the DDS `Waitset::wait()` call is non-blocking (returns without sleeping). When the timeout parameter is a non-zero duration, it sleeps for the specified duration (100 milliseconds in the example below).

dirk-thomas replied:
Ok, I was confused by the location of `blocking_timeout`.

dirk-thomas commented on test_communication/test/test_timeout_subscriber.cpp:
```diff
@@ -0,0 +1,156 @@
+    auto blocking_end = std::chrono::steady_clock::now();
+    auto blocking_diff = blocking_end - blocking_start;
+    if (blocking_diff > blocking_timeout + tolerance) {
+      fprintf(stderr, "Tolerance exceeded for 1ms wait\n");
```
Why `1ms wait`?

dirk-thomas commented on test_communication/test/test_timeout_subscriber.py.in:
```diff
@@ -0,0 +1,25 @@
+
+    assert rc == 0, \
+        "The launch file failed with exit code '" + str(rc) + "'. " \
+        'May be the subscriber did not receive any messages?'
```
That message seems to make no sense.

dirk-thomas commented on test_communication/test/test_timeout_subscriber.cpp:
```diff
@@ -0,0 +1,156 @@
+  }
+
+  size_t num_cycles = 5; //TODO: args
+  auto blocking_timeout = std::chrono::milliseconds(100);
```
This variable should be moved to line ~140 since it is not used anywhere before.

jacquelinekay commented:
I reduced the test quite a lot since I probably won't use the extra functionality copied from `test_subscriber`, and clarified/corrected some print statements.

dirk-thomas commented on test_communication/test/test_timeout_subscriber.cpp:
```diff
@@ -0,0 +1,74 @@
+
+#include "message_fixtures.hpp"
+
+int retcode;
```
Uninitialized variable. Also avoiding global variables would be better.

jacquelinekay replied:
Can I just throw a runtime error in the callback instead, since it's not expecting to receive a message?

dirk-thomas replied:
Sure.

dirk-thomas commented on test_communication/test/test_timeout_subscriber.cpp:
```diff
@@ -0,0 +1,74 @@
+
+#include <rclcpp/rclcpp.hpp>
+
+#include "message_fixtures.hpp"
```
Unused include.

dirk-thomas commented:
This should be moved to the `test_rclcpp` package as suggested here: https://github.com/ros2/system_tests/pull/20#issuecomment-126080395

jacquelinekay commented:
should I convert this test to a gtest to be consistent with the other test in `test_rclcpp` or is it fine to keep it as a nosetest?

dirk-thomas commented:
I would say keeping it as a nosetest is perfectly fine.

dirk-thomas commented:
Since the test only uses a single executable a GTest seems to be less overhead then a nosetest with a launch file.

esteve commented:
+1

wjwwood commented:
+1

tfoote commented:
Since there's not multi-node testing a gtest would be better. It's much less overhead setting up and can support much more precise testing.

dirk-thomas commented:
@jacquelinekay You have merged the other related PRs but not this one. Currently all platforms fail to build:
http://ci.ros2.org/job/ros2_batch_ci_linux/148/
http://ci.ros2.org/job/ros2_batch_ci_osx/76/
http://ci.ros2.org/job/ros2_batch_ci_windows/104/

Please fix it.

jacquelinekay commented:
I've converted the test to a Gtest and rebased. This pull request will fix the broken build, so assuming @dirk-thomas and @tfoote find these changes acceptable, I will merge after CI comes out clean. Sound good?

http://ci.ros2.org/job/ros2_batch_ci_linux/149/
http://ci.ros2.org/job/ros2_batch_ci_osx/77/
http://ci.ros2.org/job/ros2_batch_ci_windows/105/

dirk-thomas commented:
The commit log and diff contains entries which are already on master. So something is wrong with the rebase of this branch.

dirk-thomas commented:
The Windows build failed with:

    Cannot run when setup is in progress.

May be @j-rivero is currently installing the released version of VS?

j-rivero commented:
Ouch yes, sorry, I started but did not finish the process, mainly because of [this question](https://github.com/ros2/ros2/issues/72#issuecomment-127825345).

jacquelinekay commented:
Did not pull before rebasing, should look better now.

dirk-thomas commented on test_rclcpp/test/test_subscription.cpp:
```diff
@@ -22,7 +22,7 @@
 
 #ifdef RMW_IMPLEMENTATION
-# define CLASSNAME_(NAME, SUFFIX) NAME ## __ ## SUFFIX
+# define CLASSNAME_(NAME, SUFFIX) NAME##__##SUFFIX
```
Please don't change this. You likely don't have an up-to-date version of the ament linter repo.

dirk-thomas commented on test_rclcpp/CMakeLists.txt:
```diff
@@ -69,6 +69,26 @@ if(AMENT_ENABLE_TESTING)
+      target_link_libraries(gtest_timeout_subscriber__${middleware_impl}
+        ${_AMENT_EXPORT_ABSOLUTE_LIBRARIES}
+        ${_AMENT_EXPORT_LIBRARY_TARGETS}
+        ${GTEST_MAIN_LIBRARIES})
```
You should not use `GTEST_MAIN_LIBRARIES` in combination with `ament_add_gtest`.

jacquelinekay replied:
I copied this block from the above test (test_timer). why is it being used
there?

On Wed, Aug 5, 2015 at 8:57 AM, Dirk Thomas <notifications@github.com>
wrote:

> In test_rclcpp/CMakeLists.txt
> <https://github.com/ros2/system_tests/pull/19#discussion_r36318796>:
>
> > @@ -69,6 +69,26 @@ if(AMENT_ENABLE_TESTING)
> >          "${middleware_impl}"
> >          "rclcpp")
> >      endif()
> > +
> > +    # test_timeout_subscriber
> > +    ament_add_gtest(
> > +      gtest_timeout_subscriber__${middleware_impl}
> > +      "test/test_timeout_subscriber.cpp"
> > +      TIMEOUT 30
> > +    )
> > +    if(TARGET gtest_timeout_subscriber__${middleware_impl})
> > +      target_link_libraries(gtest_timeout_subscriber__${middleware_impl}
> > +        ${_AMENT_EXPORT_ABSOLUTE_LIBRARIES}
> > +        ${_AMENT_EXPORT_LIBRARY_TARGETS}
> > +        ${GTEST_MAIN_LIBRARIES})
>
> You should not use GTEST_MAIN_LIBRARIES in combination with
> ament_add_gtest.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ros2/system_tests/pull/19/files#r36318796>.
>

dirk-thomas replied:
That is wrong for `test_timer` then (removed in ). But `test_timer.cpp` also does not have a main itself.

dirk-thomas commented on test_rclcpp/CMakeLists.txt:
```diff
@@ -69,6 +69,26 @@ if(AMENT_ENABLE_TESTING)
+      gtest_timeout_subscriber__${middleware_impl}
+      "test/test_timeout_subscriber.cpp"
+      TIMEOUT 30
+    )
```
The source file `test_timeout_subscriber.cpp` has a main function and you also link against the default `main` of gtest. This will likely not work.

jacquelinekay replied:
this should be fixed now

jacquelinekay commented:
1. Will leaving out the target_compile_definitions cause the jobs that I just launched to fail? I tried individual executable on my machine and it worked without that line, but it may not work on Jenkins. If so, I will relaunch.
2. What are we going to do about the broken Windows build? Do you prefer we wait until it's fixed before this is merged?

dirk-thomas commented:
Without the definition the three result files will have the same name and therefore don't tell the user which exact test is running. Leaving it out won't affect the result of the build - only the generated test result files. You can keep the current CI jobs running.

We still need to merge this asap since currently nobody can build the master branches. As soon as the Windows build works again you should check for regressions and address them after the fact.

jacquelinekay commented:
http://ci.ros2.org/job/ros2_batch_ci_linux/150/
http://ci.ros2.org/job/ros2_batch_ci_osx/78/

j-rivero commented:
I will relaunch the job and post the results or feedback here as soon as the windows node is back.

jacquelinekay commented:
Linux and OSX are cleaning. I'm merging.
